src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34470,6 +34470,232 @@ static void F(A<object?>[] c)
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable<object>)t1").WithLocation(13, 27),
+                // (19,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var w in (IEnumerable?)t2) // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable?)t2").WithLocation(19, 27),
+                // (22,27): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         foreach (var v in (IEnumerable)t2) // 5
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(IEnumerable)t2").WithLocation(22, 27),
+                // (22,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var v in (IEnumerable)t2) // 5
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable)t2").WithLocation(22, 27));
+        }
+
+        [WorkItem(23493, "https://github.com/dotnet/roslyn/issues/23493")]
+        [Fact]
+        public void ForEach_14()
+        {
+            var source =
+@"using System.Collections;
+using System.Collections.Generic;
+class C
+{
+    static void F1<T>(T t1) where T : IEnumerable?
+    {
+        foreach (var x in t1) // 1
+        {
+        }
+        foreach (var w in (IEnumerable?)t1) // 2
+        {
+        }
+        foreach (var v in (IEnumerable)t1) // 3
+        {
+        }
+    }
+    static void F2<T>(T t2)
+    {
+        foreach (var y in (IEnumerable<object>?)t2) // 4
+        {
+        }
+        foreach (var z in (IEnumerable<object>)t2) // 5
+        {
+        }
+    }
+}";
+            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
+            comp.VerifyDiagnostics(
+                // (7,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var x in t1) // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(7, 27),
+                // (10,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var w in (IEnumerable?)t1) // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable?)t1").WithLocation(10, 27),
+                // (13,27): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(IEnumerable)t1").WithLocation(13, 27),
+                // (13,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable)t1").WithLocation(13, 27));
+        }
+
+        [Fact]
+        public void ForEach_15()
```

Comment:
>ForEach_15 [](start = 20, length = 10)

Some more test ideas:
- `MyEnumerator GetEnumerator()` where `MyEnumerator.Current` returns a nullable type, but the element type of the `foreach` is non-nullable
- `MyEnumerator GetEnumerator()` where `MyEnumerator.Current` returns a non-nullable type which has a user-defined conversion to the non-nullable element type of the `foreach`, but that user-defined conversion returns a nullable

Reply:
`ForEach_02` tests `object? Current { get; }`. Added a test with a user-defined conversion.

---
In reply to: [214110121](https://github.com/dotnet/roslyn/pull/29587#discussion_r214110121) [](ancestors = 214110121)